src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -333,44 +288,37 @@ public static void WriteTo(TextWriter writer, HtmlEncoder encoder, object value)
             var htmlContent = value as IHtmlContent;
             if (htmlContent != null)
             {
-                var htmlTextWriter = writer as HtmlTextWriter;
-                if (htmlTextWriter == null)
-                {
-                    htmlContent.WriteTo(writer, encoder);
-                }
-                else
-                {
-                    // This special case allows us to keep buffering as IHtmlContent until we get to the 'final'
-                    // TextWriter.
-                    htmlTextWriter.Write(htmlContent);
-                }
 
-                return;
+                // This special case allows us to keep buffering as IHtmlContent until we get to the 'final'
+                // IHtmlContentBuilder.
+                contentBuilder.Append(htmlContent);
+            }
+            else if (value is HelperResult)
```

Comment:
The `HelperResult` delegate changed from `Func<TextWriter, Task>`  to `Func<IHtmlContentBuilder, Task>` - which made it difficult to continue treating it as a `IHtmlContent`.